File: parentesis/sol.py
N = 55
M = 205
t = int(input())
for _ in range(t):
    dp = [[0] * (N*M*2) for _ in range(N)]
    line = input()
    lin = iter(line.split())
    ig = int(next(lin))
    dp[0][N*M] = 1
    cnt = 0
    for c, x in zip(lin, lin):
        x = int(x)
        for v in range(N*M*2):
            p = dp[cnt][v]
            if p:
                if c == '-': p += 1
                dp[cnt + 1][v + (x if p % 2 else -x)] = max(dp[cnt + 1][v + (x if p % 2 else -x)], p)
                p -= 1
                if c != '-' or p > 1: dp[cnt + 1][v + (x if p % 2 else -x)] = max(dp[cnt + 1][v + (x if p % 2 else -x)], p)
        cnt += 1
    ans = sum(1 for i in range(N * M * 2) if dp[cnt][i] != 0)
    print(ans)

File: parentesis/sol.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 55;
const int M = 205;
const int D = M*N;

int dp[N][N*M*2]; // pos, value (offset) = parenthesis remaining

int main() {
    int t;
    cin >> t;
    string ig;
    getline(cin, ig);
    while (t--) {
        memset(dp, 0, sizeof(dp));
        string line;
        getline(cin, line);
        stringstream lin(line);
        int x;
        lin >> x;
        char c;
        dp[0][D] = 1;
        int cnt = 0;
        while (lin >> c >> x) {
            for (int v = 0; v < N*M*2; ++v) {
                int p = dp[cnt][v];
                if (p) {
                    if (c == '-') ++p;
                    dp[cnt+1][v+((p%2)?x:-x)] = max(dp[cnt+1][v+((p%2)?x:-x)], p);
                    --p;
                    if (c != '-' || p > 1) {
                        dp[cnt+1][v+((p%2)?x:-x)] = max(dp[cnt+1][v+((p%2)?x:-x)], p);
                    }
                }
            }
            ++cnt;
        }
        int ans = 0;
        for (int i = 0; i < N*M*2; ++i) ans += (dp[cnt][i] != 0);
        cout << ans << '\n'; 
    }
}